pkg/util/util.go

```diff
@@ -234,61 +244,84 @@ func ParseClusterYaml(file string) (*clusterv1.Cluster, error) {
 }
 
 // ParseMachinesYaml extracts machine objects from a file.
-func ParseMachinesYaml(file string) ([]*clusterv1.Machine, error) {
+func ParseMachinesYaml(file string) ([]*clusterv1.Machine, []*clusterv1.MachineDeployment, error) {
 	reader, err := os.Open(file)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	defer reader.Close()
 
 	decoder := yaml.NewYAMLOrJSONDecoder(reader, 32)
 
 	var (
-		bytes       [][]byte
-		machineList clusterv1.MachineList
-		machine     clusterv1.Machine
-		machines    = []clusterv1.Machine{}
+		bytes              [][]byte
+		machineList        clusterv1.MachineList
+		machine            clusterv1.Machine
+		machines           = []clusterv1.Machine{}
+		machineDeployment  clusterv1.MachineDeployment
+		machineDeployments = []clusterv1.MachineDeployment{}
 	)
 
 	// TODO: use the universal decoder instead of doing this.
 	if bytes, err = decodeClusterV1Kinds(decoder, "MachineList"); err != nil {
 		if isMissingKind(err) {
 			err = errors.New(MachineListFormatDeprecationMessage)
 		}
-		return nil, err
+		return nil, nil, err
 	}
 
 	// TODO: this is O(n^2) and must be optimized
 	for _, ml := range bytes {
 		if err := json.Unmarshal(ml, &machineList); err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 		for _, machine := range machineList.Items {
 			if machine.APIVersion == "" || machine.Kind == "" {
-				return nil, errors.New(MachineListFormatDeprecationMessage)
+				return nil, nil, errors.New(MachineListFormatDeprecationMessage)
 			}
 			machines = append(machines, machine)
 		}
 	}
 
 	// reset reader to search for discrete Machine definitions
 	if _, err := reader.Seek(0, 0); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	if bytes, err = decodeClusterV1Kinds(decoder, "Machine"); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	for _, m := range bytes {
 		if err := json.Unmarshal(m, &machine); err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 		machines = append(machines, machine)
 	}
 
-	return MachineP(machines), nil
+	// reset reader to search for discrete MachineDeployment definitions
+	if _, err := reader.Seek(0, 0); err != nil {
+		return nil, nil, err
+	}
+
+	if bytes, err = decodeClusterV1Kinds(decoder, "MachineDeployment"); err != nil {
+		return nil, nil, err
+	}
+
+	for _, m := range bytes {
+		if err := json.Unmarshal(m, &machineDeployment); err != nil {
+			return nil, nil, err
+		}
+
+		// Ignore MachineDeployments for controlplane machines
+		// TODO(rudoi): warn when one of these is detected?
+		if !IsControlPlaneMachine(machineDeployment.Spec.Template.Spec) {
```

Comment:
As a user of this function I would expect it to return an unfiltered list.

Reply:
Perhaps we can create an `ExtractControlPlaneMachineDeployment` function equivalent to: https://github.com/newrelic-forks/cluster-api/blob/0b43482f7e7fb6a7da30e972d02594f4ed63a376/cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go#L1032
This might be useful to support an HA control plane in the future.